compiler/plugins/input/StableHLO/stablehlo-iree/Conversion/Preprocessing/LimitScatterBounds.cpp

```diff
@@ -0,0 +1,281 @@
+
+  // We subtract the width of the update for each index. For inserted dimensions
+  // this is 1.
+  llvm::SmallVector<Value> windowBounds;
+  {
+    int j = 0;
+    for (auto it : llvm::enumerate(isUpdatedWindowDim)) {
+      Value dstSz = b.create<tensor::DimOp>(dest, it.index());
+      if (!it.value()) {
+        windowBounds.push_back(
+            b.create<arith::SubIOp>(b.getIndexType(), dstSz, oneIdx));
+        continue;
+      }
+
+      windowBounds.push_back(b.create<arith::SubIOp>(b.getIndexType(), dstSz,
+                                                     updateWindowShape[j]));
+      ++j;
+    }
+  }
+
+  // Reorder / select the dimensions to the index ordering:
+  llvm::SmallVector<Value> reorderedBounds;
+  for (auto dim : toOperandDims) {
+    reorderedBounds.push_back(
+        b.create<arith::IndexCastOp>(indicesETy, windowBounds[dim]));
+  }
+
+  // Combine the selected dimensions into a small vector for the bounds check:
+  Value dimLimits = b.create<tensor::FromElementsOp>(reorderedBounds);
+
+  Value bounded = b.create<tensor::EmptyOp>(indicesShape, indicesETy);
+  Value inbounds = b.create<tensor::EmptyOp>(indicesShape, b.getI1Type());
+  AffineMap boundedMap = b.getMultiDimIdentityMap(indicesTy.getRank());
+
+  llvm::SmallVector<AffineExpr> boundsExpr{
+      b.getAffineDimExpr(indicesTy.getRank() - 1)};
+  AffineMap boundsMap =
+      AffineMap::get(indicesTy.getRank(), 0, boundsExpr, b.getContext());
+
+  SmallVector<utils::IteratorType> indicesIter(indicesTy.getRank(),
+                                               utils::IteratorType::parallel);
+  auto applyLimits = b.create<linalg::GenericOp>(
+      TypeRange{indicesTy, inbounds.getType()}, ValueRange{indices, dimLimits},
+      ValueRange{bounded, inbounds},
+      ArrayRef<AffineMap>{boundedMap, boundsMap, boundedMap, boundedMap},
+      indicesIter, [](OpBuilder &bb, Location loc, ValueRange args) {
+        ImplicitLocOpBuilder b(loc, bb);
+        Value zero =
+            b.create<arith::ConstantOp>(b.getZeroAttr(args[0].getType()));
+        Value lower =
+            b.create<arith::CmpIOp>(arith::CmpIPredicate::sge, args[0], zero);
+        Value upper = b.create<arith::CmpIOp>(arith::CmpIPredicate::sle,
+                                              args[0], args[1]);
+        Value within = b.create<arith::AndIOp>(lower, upper);
+        Value sel = b.create<arith::SelectOp>(within, args[0], zero);
+        b.create<linalg::YieldOp>(ValueRange{sel, within});
+      });
+
+  bounded = applyLimits.getResult(0);
+  inbounds = applyLimits.getResult(1);
+
+  // If the index is not implicit we need to reduce away the index vector dim:
+  if (indexVectorDim < indicesTy.getRank()) {
+    ShapedType inTy = cast<ShapedType>(inbounds.getType());
+
+    llvm::SmallVector<AffineExpr> inExprs;
+    llvm::SmallVector<AffineExpr> outExprs;
+    llvm::SmallVector<int64_t> outShape;
+    for (int i = 0; i < inTy.getRank(); ++i) {
```

Comment:
Could we do range based on inTy dims? (seems i's only used for that)

Reply:
Technically we need the index as well but I changed to an `llvm::enumerate` for ease.